File: test/sslserver.c
#include <stdio.h>
#include "chuck.h"

chk_event_loop *loop;

chk_stream_socket_option option = {
	.recv_buffer_size = 4096,
	.decoder = NULL,
};

void data_event_cb(chk_stream_socket *s,chk_bytebuffer *data,int32_t error) {
	if(data){
		printf("recv request:%s\n",data->head->data);
		chk_stream_socket_send(s,chk_bytebuffer_clone(data),NULL,NULL);
	}
	else{
		printf("socket close:%d\n",error);
	}
}

void accept_cb_ssl(chk_acceptor *a,int32_t fd,chk_sockaddr *addr,void *ud,int32_t err) {
	
	if(fd >= 0) {

		chk_stream_socket *s = chk_stream_socket_new(fd,&option);

		SSL_CTX *ssl_ctx = chk_acceptor_get_ssl_ctx(a);

		if(0 == chk_ssl_accept(s,ssl_ctx)){
			chk_loop_add_handle(loop,(chk_handle*)s,data_event_cb);
		}
		else {
			chk_stream_socket_close(s,0);
			printf("chk_ssl_accept error\n");
		}
	}
	else {
		printf("accept error:%d\n",err);
	}
}

static void signal_int(void *ud) {
	printf("signal_int\n");
	chk_loop_end(loop);
}

int main(int argc,char **argv) {

	const char *certificate = "./test/cacert.pem";
	const char *privatekey = "./test/privkey.pem";

	signal(SIGPIPE,SIG_IGN);
	loop = chk_loop_new();

	if(0 != chk_watch_signal(loop,SIGINT,signal_int,NULL,NULL)) {
		printf("watch signal failed\n");
		return 0;
	}

    SSL_CTX *ctx = SSL_CTX_new(SSLv23_server_method());
    /* 也可以用 SSLv2_server_method() 或 SSLv3_server_method() 单独表示 V2 或 V3标准 */
    if (ctx == NULL) {
        ERR_print_errors_fp(stdout);
        return 0;
    }
    /* 载入用户的数字证书， 此证书用来发送给客户端。 证书里包含有公钥 */
    if (SSL_CTX_use_certificate_file(ctx,certificate, SSL_FILETYPE_PEM) <= 0) {
        ERR_print_errors_fp(stdout);
        return 0;
    }
    /* 载入用户私钥 */
    if (SSL_CTX_use_PrivateKey_file(ctx, privatekey, SSL_FILETYPE_PEM) <= 0) {
        ERR_print_errors_fp(stdout);
        return 0;
    }
    /* 检查用户私钥是否正确 */
    if (!SSL_CTX_check_private_key(ctx)) {
        ERR_print_errors_fp(stdout);
        return 0;
    }	

	if(!chk_ssl_listen_tcp_ip4(loop,argv[1],atoi(argv[2]),ctx,accept_cb_ssl,NULL)){
		printf("server start error\n");
	}
	else{
		CHK_SYSLOG(LOG_INFO,"server start");
		chk_loop_run(loop);
	}

	chk_loop_del(loop);
	return 0;
}

File: test/lua/sslserver.lua
package.path = './lib/?.lua;'
package.cpath = './lib/?.so;'

local chuck = require("chuck")
local socket = chuck.socket
local packet = chuck.packet
local ssl = chuck.ssl

local event_loop = chuck.event_loop.New()


local ssl_ctx

local function init_ssl()
	local _ssl_ctx = ssl.SSL_CTX_new("SSLv23_server_method")

	if not _ssl_ctx then
		return nil 
	end

	local err = ssl.SSL_CTX_use_certificate_file(_ssl_ctx,"./test/cacert.pem")

	if err then
		print("SSL_CTX_use_certificate_file failed:" .. err)
		return nil
	end

	err = ssl.SSL_CTX_use_PrivateKey_file(_ssl_ctx,"./test/privkey.pem")

	if err then
		print("SSL_CTX_use_PrivateKey_file failed:" .. err)		
		return nil
	end

	err = ssl.SSL_CTX_check_private_key(_ssl_ctx)

	if err then
		print("SSL_CTX_check_private_key failed:" .. err)				
		return nil
	end

	return _ssl_ctx

end


local ssl_ctx = init_ssl()

local serverAddr = socket.addr(socket.AF_INET,"127.0.0.1",8010)

if ssl_ctx then 
	local server = socket.stream.listen(event_loop,serverAddr,function (fd,err)
		if err then
			print("err:" .. err)
			return
		end

		local conn = socket.stream.socket(fd,4096,packet.Decoder(65536))
		if conn then
			local err = ssl.SSL_accept(conn,ssl_ctx)
			if err then
				print("SSL_accept error:" .. err)
				conn:Close()
				return
			end

			print("new client")			

			conn:Start(event_loop,function (data)
				if data then
					--conn:Close()
					conn:Send(data)
				else
					print("client disconnected") 
					conn:Close()
				end
			end)
		end
	end)

	print("server run")

	if server then
		event_loop:WatchSignal(chuck.signal.SIGINT,function()
			print("recv SIGINT stop server")
			event_loop:Stop()
		end)
		event_loop:Run()
	end

else
	print("ssl init error")
end